### ISSUE TITLE
Hello, when I set allowFormAuthenticationForClients in the AuthorizationServer configuration, the customized AuthenticationEntryPoint will not be available.

### ISSUE THREAD

**Issue by g-u-a-n** [spring-projects/spring-security-oauth#1664]
I am using the version of spring-security-oauth2 2.3.5.RELEASE

When I configure the AuthorizationServer, enable **allowFormAuthenticationForClients** and customize the `AuthenticationEntryPoint`, then the `AuthenticationEntryPoint` will not work.

```
    @Configuration
    @EnableAuthorizationServer
    protected static class AuthorizationServerConfiguration extends AuthorizationServerConfigurerAdapter {
        ...
        @Override
        public void configure(AuthorizationServerSecurityConfigurer oauthServer) {
            MyAuthenticationEntryPoint myAuthenticationEntryPoint = new MyAuthenticationEntryPoint();
            myAuthenticationEntryPoint.setRealmName("oauth2/client");
            myAuthenticationEntryPoint.setExceptionTranslator(new MyWebResponseExceptionTranslator());
            oauthServer.authenticationEntryPoint(myAuthenticationEntryPoint).allowFormAuthenticationForClients();
        }
        ...
    }
```

I want to modify the exception response content through the `AuthenticationEntryPoint.setExceptionTranslator`.

`{"error":"invalid_client","error_description":"Bad client credentials"}`

I found that in the `AuthorizationServerSecurityConfigurer` `clientCredentialsTokenEndpointFilter`, an `AuthenticationEntryPoint` is hardcoded instead of my custom `AuthenticationEntryPoint`.

`org.springframework.security.oauth2.config.annotation.web.configurers.AuthorizationServerSecurityConfigurer`:

```
	@Override
	public void configure(HttpSecurity http) throws Exception {
		
		// ensure this is initialized
		...
		if (allowFormAuthenticationForClients) {
			clientCredentialsTokenEndpointFilter(http);
		}
		...
	}
	...
	private ClientCredentialsTokenEndpointFilter clientCredentialsTokenEndpointFilter(HttpSecurity http) {
		ClientCredentialsTokenEndpointFilter clientCredentialsTokenEndpointFilter = new ClientCredentialsTokenEndpointFilter(
				frameworkEndpointHandlerMapping().getServletPath("/oauth/token"));
		...
		OAuth2AuthenticationEntryPoint authenticationEntryPoint = new OAuth2AuthenticationEntryPoint();
		...
		clientCredentialsTokenEndpointFilter.setAuthenticationEntryPoint(authenticationEntryPoint);
		...
	}
```

Is this a bug? How can I solve this problem?
Sorry, my English is not very good. If you are offended by me, please forgive me!

**Comment by yuerzm** [comment]
When I disable `allowFormAuthenticationForClients`, I cannot customize and inject it into the filter.
When I enabled `allowFormAuthenticationForClients`, I have the same problem.
And I hope I can customize `authenticationEntryPoint` after using it.

**Comment by mostermsater** [comment]
like this,it's work
`@Configuration
public class BasicConfig {

	@Autowired
	@Qualifier("springSecurityFilterChain")
	private Filter springSecurityFilterChain;
	
	@Autowired
	private CustomClientAuthenticationEntryPoint customClientAuthenticationEntryPoint;
	
	@PostConstruct
	public void getFilters() {
	    FilterChainProxy filterChainProxy = (FilterChainProxy) springSecurityFilterChain;
	    List<SecurityFilterChain> list = filterChainProxy.getFilterChains();
	    list.stream()
	      .flatMap(chain -> chain.getFilters().stream()) 
	      .forEach(filter -> {
	    	  if(filter instanceof ClientCredentialsTokenEndpointFilter) {
	    		 ((ClientCredentialsTokenEndpointFilter)filter).setAuthenticationEntryPoint(customClientAuthenticationEntryPoint);
	    	  }
	      } );
	}
}`
[https://www.baeldung.com/spring-security-registered-filters](url)

**Comment by g-u-a-n** [comment]
> like this,it's work
> 
> ```
> @Autowired
> @Qualifier("springSecurityFilterChain")
> private Filter springSecurityFilterChain;
> 
> @Autowired
> private CustomClientAuthenticationEntryPoint customClientAuthenticationEntryPoint;
> 
> @PostConstruct
> public void getFilters() {
>     FilterChainProxy filterChainProxy = (FilterChainProxy) springSecurityFilterChain;
>     List<SecurityFilterChain> list = filterChainProxy.getFilterChains();
>     list.stream()
>       .flatMap(chain -> chain.getFilters().stream()) 
>       .forEach(filter -> {
>     	  if(filter instanceof ClientCredentialsTokenEndpointFilter) {
>     		 ((ClientCredentialsTokenEndpointFilter)filter).setAuthenticationEntryPoint(customClientAuthenticationEntryPoint);
>     	  }
>       } );
> }
> ```
> 
> [https://www.baeldung.com/spring-security-registered-filters](url)

Hello, friend, thank you very much for your solution, thank you!

**Comment by g-u-a-n** [comment]
You can refer to the solution provided by this friend and hope to help you. [https://github.com/spring-projects/spring-security-oauth/issues/1664#issuecomment-487855841](url)

**Comment by yuerzm** [comment]
@g-u-a-n @mostermsater Thank you for your solution!
Injection is possible, but the scenarios are limited. When you have multiple `springSecurityFilterChain` context, differences were seen running effect as expected. Also provides a method of using the same package name and class name rewrite `ClientCredentialsTokenEndpointFilter`.
I would prefer that the spring team do some extension support for this. This may not be the best approach.

**Comment by g-u-a-n** [comment]
> @g-u-a-n @mostermsater Thank you for your solution!
> Injection is possible, but the scenarios are limited. When you have multiple `springSecurityFilterChain` context, differences were seen running effect as expected. Also provides a method of using the same package name and class name rewrite `ClientCredentialsTokenEndpointFilter`.
> I would prefer that the spring team do some extension support for this. This may not be the best approach.

I agree with your thoughts, I also hope that the spring team can solve this problem from the root cause.
